src/main/java/com/aws/greengrass/util/SwappingFileWriter.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.util;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+
+/**
+ * A writer which utilizes two in-sync files and swaps pointers on every write such that the "read" file is never
+ * corrupt.
+ */
+public final class SwappingFileWriter extends Writer {
+    private final Path readFilePath;
+    private final Path writeFilePath;
+    private final Path swapFilePath;
+    private BufferedWriter writeFile;
+    private BufferedWriter readFile;
+    private boolean closed;
+
+    /**
+     * Creates a new instance of SwappingFileWriter.
+     *
+     * @param output path of the "read" file
+     * @throws IOException when creating paths, writers, or copying fails
+     */
+    public SwappingFileWriter(Path output) throws IOException {
+        super();
+        readFilePath = output;
+        writeFilePath = getWriteFile(output);
+        swapFilePath = getSwappingFile(output);
+        Utils.createPaths(readFilePath.getParent());
+
+        // Cleanup any old write file if it exists
+        Files.deleteIfExists(writeFilePath);
+        // Sync read and write files if the read file exists
+        if (Files.exists(readFilePath)) {
+            Files.copy(readFilePath, writeFilePath, StandardCopyOption.REPLACE_EXISTING);
+        }
+
+        // Now we know that the read and write files are exactly the same so we can safely begin to simply append
+        // to them without any additional, larger, copies.
+
+        readFile = Files.newBufferedWriter(readFilePath, StandardOpenOption.CREATE,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when Files.newBufferedWriter throws an exception, the *SwappingFileWriter* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *readFile* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).